[B] Add missing entity effects. Adds BUKKIT-3311

LordRalex opened the pull request:
**Issue**

Within the server are a few different entity effects that an entity will show. The current status of the enum only provides access to a few of them. There are still a few effects that a server can send that is not accessible with the enum as it is.

**Justification**

This adds the remaining entity effects that a server can send to the client.

**PR Breakdown**

This adds the remaining values of enums that the entity effect can have, which are all sent by the server as a byte.

**Testing**

To test, an entity was spawned into the world and the entity was then told to play the new effects. This was done with logging of the current effect and looking to see if an effect was played. The player was also tested to test to see if the eating effect was playable, however it was not.

``` java
    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!(sender instanceof Player)) {
            return false;
        }
        //((Player) sender).playEffect(EntityEffect.EAT_ACCEPTED);
        Entity fw = ((Player) sender).getWorld().spawn(((Player) sender).getLocation(), EntityType.valueOf(args[0].toUpperCase()).getEntityClass());

        new DelayedRunnable(fw).runTaskTimer(this, 20, 20);
        return true;
    }

    private class DelayedRunnable extends BukkitRunnable {

        private final EntityEffect[] effects = new EntityEffect[]{
            //EntityEffect.EAT_ACCEPTED,
            EntityEffect.FIREWORK_EXPLODE,
            EntityEffect.IRON_GOLEM_ROSE,
            EntityEffect.VILLAGER_ANGRY,
            EntityEffect.VILLAGER_HAPPY,
            EntityEffect.VILLAGER_HEART,
            EntityEffect.WITCH_MAGIC,
            EntityEffect.ZOMBIE_TRANSFORM
        };
        Entity fw;
        int i = 0;

        public DelayedRunnable(Entity in) {
            super();
            fw = in;
        }

        @Override
        public void run() {
            if (i >= effects.length) {
                if (fw != null) {
                    fw.remove();
                }
                fw = null;
                this.cancel();
                return;
            } else {
                Bukkit.broadcastMessage("Playing: " + effects[i].toString());
                fw.playEffect(effects[i]);
                i++;
            }
        }
    }

```

Files listed contain the fixes to the issues using them for testing: https://github.com/Bukkit/Bukkit/pull/907#issuecomment-24787334
Source: https://dl.dropboxusercontent.com/u/53849969/Bukkit-PRs/entityeffect/src.zip
Binary: https://dl.dropboxusercontent.com/u/53849969/Bukkit-PRs/entityeffect/TestPlugin-1.0-SNAPSHOT.jar

**Relevant PRs**

https://github.com/Bukkit/Bukkit/pull/904 - Initial PR that led to this one, also included a enum this one adds

**JIRA Ticket**

https://bukkit.atlassian.net/browse/BUKKIT-3311

gravitylow commented:
Have you tested to ensure all these can play properly?

LordRalex commented:
They have now been tested and the results reflected by the javadocs. The eating effect is apparently not playable from the server from the testing I have did.

gravitylow commented:
You can remove the "without client-mods" part. We don't support client mods and so those notes don't need to be on our site.

LordRalex commented:
Those were added as they mimic the existing javadocs for such effects. Should the other javadocs be updated as well because of that, or should they remain the same as each other.

Xarinor commented:
I like it - sure you've got all of the playables?

LordRalex commented:
Those are all of the ones that the server will send based on the Entity Effect related packet. There is one other, the eating, however that has no effect when forced to used, so rather than include it and have a "broken" effect, I chose to not include it. Otherwise, this is every (or should be, based on the packet used) effect that is known and can be played with playEffect.

riking commented:
What if you set their item-in-hand to something consumable then tried to play the "eating" effect?

LordRalex commented:
I had tested using the eating effect on a player who had food and looked to see if either that player or others watching him could see it. In that, neither client saw the eating effect at all. I also tested if they did not have food, and got the same result

Although since the server does permit this action, I now will be looking to see what let's clients show this effect, as to see if it is just forcing it to be the issue, or if they do not use this at all.

LordRalex commented:
Looks like the enum for eating is not even used: https://github.com/Bukkit/CraftBukkit/blob/master/src/main/java/net/minecraft/server/EntityHuman.java#L224
That is how it does feeding for players on the server, and just spawns particles with it.

turt2live commented:
Here's the question I have: Does this need a CB side? Or does it work as expected as-is?

Edit: And the title should be "Add missing effects" or something along those lines

turt2live commented on src/main/java/org/bukkit/EntityEffect.java:
```diff
@@ -45,7 +45,56 @@
+    /**
+     * When a firework explodes.
+     * <p>
+     * This will be ignored if the entity is not a firework.
```
Instead of ignored for all of these cases I would put "This does not emit an effect if the entity..."

LordRalex replied:
I was just using the same wording from the other javadocs for those effects, but that does make more sense, although "play" is probably more accurate

turt2live replied:
Very true. Don't go modifying the existing JavaDocs (that's a job for the JD repo) though, just the ones you added :)

LordRalex commented:
This does not need a CB change because of how the effects worked inside. It uses the effects as a byte value, which is sent to the client to be handled. I did test this on Craftbukkit with just this added code and using the test plugin above, and all the effects had played as expected.

turt2live commented:
Perfect, I'll test this either some time today or over the next few days. If I don't respond saying "It works!" or "It doesn't work" by say... Wednesday, ping me.

turt2live commented:
For ref: I tested this, it works.

Java 6 compatible testing materials: [Source + Jar + Plugin.yml + Instructions](https://drive.google.com/folderview?id=0B6JIYSBkLkAnLTktQS1HNnhiQlU&usp=sharing&tid=0B6JIYSBkLkAnTTZYd3Z4ZTMzV3c)

Note: The uploaded test plugin is a slightly modified version of the test plugin supplied in the description of this PR. All I did was recompile it in Java 6 and remove a line that shouldn't be there.

EDIT: This can cause errors on the server, you have been warned. /stop the server to fix it. (Errors produced: http://pastebin.com/bWtrMTVD)

riking commented:
> EDIT: This can cause errors on the server, you have been warned. /stop the server to fix it. (Errors produced: http://pastebin.com/bWtrMTVD)

Looks like incorrect use of the scheduler, mate. Does @LordRalex need to fix it, or is this good enough?

LordRalex commented:
The files I have in the top have the corrected versions, and informed @turt2live of the fix for it when he was testing it.

LordRalex committed:
Add remaining entity effects. Adds BUKKIT-3311
Some of the entity effects that exist within Minecraft were not accessible by this enum. This adds the missing effects that the server can send to a client.

turt2live commented:
Pulled, thanks!

https://github.com/Bukkit/Bukkit/commit/e802978b3c981223563f9cb3ab7e38ae01e48103
